### ISSUE TITLE
[Accuracy] Gelu has bad accuracy

### ISSUE THREAD

**Issue by sgligorijevicTT** [tenstorrent/tt-xla#618]
Flax's implementation of Gelu in exact mode(approximate=False) decomposes to a ~100 op SHLO graph or ~150 op TTIR graph, which performs many numerically unstable operations and introduces huge inaccuracies and even infinities.

Implementation appears to be performing a piecewise approximation with a bunch of WhereOp-s at the end. Right now this is a problem in cases where infs appear. Currently I'm investigating if a fixed `WhereOp` would lead to much better accuray(it is plausible that in all cases where numerics explode, that branch isn't taken)

This was determined while debugging Bert

Longer term we should also chase down inaccuracy in every single op part of this graph.
Another longer term goal would be to fuse operations like these.
Attaching StableHLO for reference:
```
module @jit_gelu attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<32x32xf32> {mhlo.sharding = "{replicated}"}) -> (tensor<32x32xf32> {jax.result_info = ""}) {
    %cst = stablehlo.constant dense<1.12837911> : tensor<32x32xf32>
    %cst_0 = stablehlo.constant dense<-0.37612626> : tensor<32x32xf32>
    %cst_1 = stablehlo.constant dense<0.112835854> : tensor<32x32xf32>
    %cst_2 = stablehlo.constant dense<-0.0268538129> : tensor<32x32xf32>
    %cst_3 = stablehlo.constant dense<0.00518832775> : tensor<32x32xf32>
    %cst_4 = stablehlo.constant dense<-8.0101937E-4> : tensor<32x32xf32>
    %cst_5 = stablehlo.constant dense<7.85386146E-5> : tensor<32x32xf32>
    %cst_6 = stablehlo.constant dense<0.000000e+00> : tensor<32x32xf32>
    %cst_7 = stablehlo.constant dense<-88.7228394> : tensor<32x32xf32>
    %cst_8 = stablehlo.constant dense<0.564189494> : tensor<32x32xf32>
    %cst_9 = stablehlo.constant dense<-0.282076746> : tensor<32x32xf32>
    %cst_10 = stablehlo.constant dense<0.42184633> : tensor<32x32xf32>
    %cst_11 = stablehlo.constant dense<-1.01526523> : tensor<32x32xf32>
    %cst_12 = stablehlo.constant dense<2.92101908> : tensor<32x32xf32>
    %cst_13 = stablehlo.constant dense<-7.49551868> : tensor<32x32xf32>
    %cst_14 = stablehlo.constant dense<1.297720e+01> : tensor<32x32xf32>
    %cst_15 = stablehlo.constant dense<-10.477664> : tensor<32x32xf32>
    %cst_16 = stablehlo.constant dense<0.563825965> : tensor<32x32xf32>
    %cst_17 = stablehlo.constant dense<-0.274112701> : tensor<32x32xf32>
    %cst_18 = stablehlo.constant dense<3.404880e-01> : tensor<32x32xf32>
    %cst_19 = stablehlo.constant dense<-0.494451523> : tensor<32x32xf32>
    %cst_20 = stablehlo.constant dense<0.621000468> : tensor<32x32xf32>
    %cst_21 = stablehlo.constant dense<-0.582473278> : tensor<32x32xf32>
    %cst_22 = stablehlo.constant dense<0.368742466> : tensor<32x32xf32>
    %cst_23 = stablehlo.constant dense<-0.138703942> : tensor<32x32xf32>
    %cst_24 = stablehlo.constant dense<2.326820e-02> : tensor<32x32xf32>
    %cst_25 = stablehlo.constant dense<2.000000e+00> : tensor<32x32xf32>
    %cst_26 = stablehlo.constant dense<1.000000e+00> : tensor<32x32xf32>
    %cst_27 = stablehlo.constant dense<0.707106769> : tensor<f32>
    %cst_28 = stablehlo.constant dense<5.000000e-01> : tensor<f32>
    %0 = stablehlo.broadcast_in_dim %cst_28, dims = [] : (tensor<f32>) -> tensor<32x32xf32>
    %1 = stablehlo.multiply %0, %arg0 : tensor<32x32xf32>
    %2 = stablehlo.negate %arg0 : tensor<32x32xf32>
    %3 = stablehlo.broadcast_in_dim %cst_27, dims = [] : (tensor<f32>) -> tensor<32x32xf32>
    %4 = stablehlo.multiply %2, %3 : tensor<32x32xf32>
    %5 = stablehlo.multiply %4, %4 : tensor<32x32xf32>
    %6 = stablehlo.negate %5 : tensor<32x32xf32>
    %7 = stablehlo.abs %4 : tensor<32x32xf32>
    %8 = stablehlo.divide %cst_26, %5 : tensor<32x32xf32>
    %9 = stablehlo.exponential %6 : tensor<32x32xf32>
    %10 = stablehlo.divide %cst_26, %7 : tensor<32x32xf32>
    %11 = stablehlo.multiply %9, %10 : tensor<32x32xf32>
    %12 = stablehlo.compare  LT, %7, %cst_25 : (tensor<32x32xf32>, tensor<32x32xf32>) -> tensor<32x32xi1>
    %13 = stablehlo.multiply %cst_24, %8 : tensor<32x32xf32>
    %14 = stablehlo.add %13, %cst_23 : tensor<32x32xf32>
    %15 = stablehlo.multiply %14, %8 : tensor<32x32xf32>
    %16 = stablehlo.add %15, %cst_22 : tensor<32x32xf32>
    %17 = stablehlo.multiply %16, %8 : tensor<32x32xf32>
    %18 = stablehlo.add %17, %cst_21 : tensor<32x32xf32>
    %19 = stablehlo.multiply %18, %8 : tensor<32x32xf32>
    %20 = stablehlo.add %19, %cst_20 : tensor<32x32xf32>
    %21 = stablehlo.multiply %20, %8 : tensor<32x32xf32>
    %22 = stablehlo.add %21, %cst_19 : tensor<32x32xf32>
    %23 = stablehlo.multiply %22, %8 : tensor<32x32xf32>
    %24 = stablehlo.add %23, %cst_18 : tensor<32x32xf32>
    %25 = stablehlo.multiply %24, %8 : tensor<32x32xf32>
    %26 = stablehlo.add %25, %cst_17 : tensor<32x32xf32>
    %27 = stablehlo.multiply %26, %8 : tensor<32x32xf32>
    %28 = stablehlo.add %27, %cst_16 : tensor<32x32xf32>
    %29 = stablehlo.multiply %cst_15, %8 : tensor<32x32xf32>
    %30 = stablehlo.add %29, %cst_14 : tensor<32x32xf32>
    %31 = stablehlo.multiply %30, %8 : tensor<32x32xf32>
    %32 = stablehlo.add %31, %cst_13 : tensor<32x32xf32>
    %33 = stablehlo.multiply %32, %8 : tensor<32x32xf32>
    %34 = stablehlo.add %33, %cst_12 : tensor<32x32xf32>
    %35 = stablehlo.multiply %34, %8 : tensor<32x32xf32>
    %36 = stablehlo.add %35, %cst_11 : tensor<32x32xf32>
    %37 = stablehlo.multiply %36, %8 : tensor<32x32xf32>
    %38 = stablehlo.add %37, %cst_10 : tensor<32x32xf32>
    %39 = stablehlo.multiply %38, %8 : tensor<32x32xf32>
    %40 = stablehlo.add %39, %cst_9 : tensor<32x32xf32>
    %41 = stablehlo.multiply %40, %8 : tensor<32x32xf32>
    %42 = stablehlo.add %41, %cst_8 : tensor<32x32xf32>
    %43 = stablehlo.select %12, %28, %42 : tensor<32x32xi1>, tensor<32x32xf32>
    %44 = stablehlo.multiply %11, %43 : tensor<32x32xf32>
    %45 = stablehlo.compare  LT, %6, %cst_7 : (tensor<32x32xf32>, tensor<32x32xf32>) -> tensor<32x32xi1>
    %46 = stablehlo.select %45, %cst_6, %44 : tensor<32x32xi1>, tensor<32x32xf32>
    %47 = stablehlo.compare  LT, %4, %cst_6 : (tensor<32x32xf32>, tensor<32x32xf32>) -> tensor<32x32xi1>
    %48 = stablehlo.subtract %cst_25, %46 : tensor<32x32xf32>
    %49 = stablehlo.select %47, %48, %46 : tensor<32x32xi1>, tensor<32x32xf32>
    %50 = stablehlo.multiply %4, %4 : tensor<32x32xf32>
    %51 = stablehlo.multiply %cst_5, %50 : tensor<32x32xf32>
    %52 = stablehlo.add %51, %cst_4 : tensor<32x32xf32>
    %53 = stablehlo.multiply %52, %50 : tensor<32x32xf32>
    %54 = stablehlo.add %53, %cst_3 : tensor<32x32xf32>
    %55 = stablehlo.multiply %54, %50 : tensor<32x32xf32>
    %56 = stablehlo.add %55, %cst_2 : tensor<32x32xf32>
    %57 = stablehlo.multiply %56, %50 : tensor<32x32xf32>
    %58 = stablehlo.add %57, %cst_1 : tensor<32x32xf32>
    %59 = stablehlo.multiply %58, %50 : tensor<32x32xf32>
    %60 = stablehlo.add %59, %cst_0 : tensor<32x32xf32>
    %61 = stablehlo.multiply %60, %50 : tensor<32x32xf32>
    %62 = stablehlo.add %61, %cst : tensor<32x32xf32>
    %63 = stablehlo.multiply %4, %62 : tensor<32x32xf32>
    %64 = stablehlo.subtract %cst_26, %63 : tensor<32x32xf32>
    %65 = stablehlo.abs %4 : tensor<32x32xf32>
    %66 = stablehlo.compare  LT, %65, %cst_26 : (tensor<32x32xf32>, tensor<32x32xf32>) -> tensor<32x32xi1>
    %67 = stablehlo.select %66, %64, %49 : tensor<32x32xi1>, tensor<32x32xf32>
    %68 = stablehlo.multiply %1, %67 : tensor<32x32xf32>
    return %68 : tensor<32x32xf32>
  }
}
```

**Comment by LPanosTT** [comment]
This branch (`lpanos/erfc_fusion`) has a pattern which will fuse the majority of this pattern. It will leave gelu represented in the form:
```
gelu(x) = (0.5*x)*(1-erf(-x*sqrt(0.5)))
```

Further work to be done to fuse the rest of it. This branch is a fork of `lpanos/erf_gelu_fusion`.
